File: day-01/js/main.ts
const input = Deno.readTextFileSync(Deno.args[0]);
const lines = input.split(/\n/);

const leftList = [];
const rightList = [];

for (const line of lines) {
	if (line === "") continue;

	const sides = line.split(/\s+/);
	leftList.push(Number(sides[0]));
	rightList.push(Number(sides[1]));
}

console.log("The total sum of the differences is", sumOfDifferences(leftList, rightList));
console.log("The total sum of the similarities is", sumOfSimilarities(leftList, rightList));

// part 1
function sumOfDifferences(leftList: number[], rightList: number[]): number {
	leftList.sort((a, b) => a - b);
	rightList.sort((a, b) => a - b);

	let differences = 0;

	for (let i = 0; i < leftList.length; i++) {
		differences += Math.abs(leftList[i] - rightList[i]);
	}

	return differences;
}

// part 2
function sumOfSimilarities(leftList: number[], rightList: number[]): number {
	// set of all unique IDs in the leftList
	const leftSet = new Set(leftList);

	// map of all unique IDs with the key = the ID itself and the value = the number of times it appeared
	const rightMap = new Map<number, number>();
	for (const x of rightList) {
		rightMap.set(x, (rightMap.get(x) ?? 0) + 1);
	}

	let similarities = 0;
	for (const id of leftSet) {
		similarities += id * (rightMap.get(id) ?? 0);
	}

	return similarities;
}

File: day-01/rust/src/main.rs
use std::collections::{HashMap, HashSet};

const INPUT: &str = include_str!("../../input.txt");

fn main() {
    let (mut left, mut right): (Vec<i32>, Vec<i32>) = INPUT
        .lines()
        .map(|line| {
            let (left, right) = line.split_once("   ").unwrap();
            (left.parse::<i32>().unwrap(), right.parse::<i32>().unwrap())
        })
        .unzip();

    println!(
        "The total sum of the differences is {}",
        sum_of_differences(&mut left, &mut right)
    );

    println!(
        "The total sum of the similarities is {}",
        sum_of_similarities(&left, &right)
    );
}

fn sum_of_differences(left: &mut [i32], right: &mut [i32]) -> i32 {
    left.sort();
    right.sort();

    left.iter()
        .zip(right.iter())
        .fold(0, |acc, (&left, &right)| acc + (left - right).abs())
}

fn sum_of_similarities(left: &[i32], right: &[i32]) -> i32 {
    // set of all unique IDs in the left list
    let left_set = left.iter().copied().collect::<HashSet<_>>();

    // map of all unique IDs with the key = the ID itself and the value = the number of times it appeared
    let right_map = right.iter().fold(
        HashMap::<i32, i32>::new(),
        |mut right_map_by_occurance, &id| {
            *right_map_by_occurance.entry(id).or_default() += 1;
            right_map_by_occurance
        },
    );

    left_set.into_iter().fold(0, |acc, id| {
        acc + id * right_map.get(&id).copied().unwrap_or_default()
    })
}
